### ISSUE TITLE
Voice VOIP garbled & slow on WAN when using MicrophoneTransmitter & MicrophoneReceiver

### ISSUE THREAD

**Issue by StokedOnIt** [Microsoft/HoloToolkit#72]
Voice VOIP garbled & slow on WAN when using MicrophoneTransmitter & MicrophoneReceiver. 
Could this just be an incorrect setting?

**Comment by PaulKemppi** [comment]
I have noticed the same problem when using WAN.

**Comment by PaulKemppi** [comment]
Apparently the problem is related to the available bandwidth. It seems that the VoIP is both sending and receiving 3-4Mbps(!) even when low quality voice option. Apparently the audio stream is not compressed at all.

**Comment by StephenHodgson** [comment]
Audio compression would definitely help. I wonder if there's already libraries for audio compression inside the toolkit that forgot to get implemented or if that is something developers will need to implement themselves.

**Comment by PaulKemppi** [comment]
Yes indeed. It seems a bit weird that the Sharing module has an external dependency called speex-1.2 which seems to be audio compression codec? Apparently it is not used.

**Comment by PaulKemppi** [comment]
In SharingService -> AudioSessionProcessorServer.h / cpp the compressor and decompressor (SpeexProcessorPtr) have been commented out. It would be nice to know the reason. We would really like to get the Sharing module working with a low(wer) bandwidth connection.

**Comment by mr0ng** [comment]
@StokedOnIt how did you get the voice chat to work? I tried attaching both the transmitter and receiver scripts to the camera (and a few other objects), but I can't seem to transmit any audio at all - both LAN and WAN. You mentioned in another tread that you had to tweak some code. Any help is appreciated!

**Comment by PaulKemppi** [comment]
Yes I got it working, but it requires a good Wi-Fi connection. You need you have Sharing prefab in both ends with IsAudioEndPoint checked. Then like you mentioned, attach MicrophoneReceiver and Transmitter components to your Camera. Additionally you need to establish connection between the endpoints (or check ConnectOnAwake from Sharing). And you have to have the Sharing server running with IP address accessible to the clients. That's all. Due to the bandwidth problems, I decided to give up the whole Sharing module and to implement my own VoIP using Speex audio compression. Not finished with that yet, but the path looks promising.

**Comment by branchbark** [comment]
Can we do voice commands if we are streaming audio? 

I'm having trouble transmitting/receiving any voice at all.  I can only hear myself when I speak however I am unable to hear others who have joined the app and they cannot hear me.

**Comment by mr0ng** [comment]
@branchbark @PaulKemppi I got the voice transmission working over WAN, but as you mentioned - extremely bad quality and unusable. Checking to see if this is still an outstanding issue and if anyone is working on a fix.

**Comment by sorianog** [comment]
Did anyone ever observe the packets being transmitted when audio is being streamed? It would be nice to see some data on if there are some important things being dropped on the network.

**Comment by mr0ng** [comment]
All - Whenever we do get a decent connection over WAN, we hear echos of ourselves. This only happens on the HoloLens and not when we connect via the Unity editor. Any suggestions on removing the echo? We've determined that it's not caused by speaker-mic feedback (we tested with earphones). Thank you! I put the microphone transmitter on the HoloLens Camera and I put the Microphone receiver on another gameobject in the scene.
